dm/docs/RFCS/20220721_continuous_data_validation.md

```diff
@@ -0,0 +1,64 @@
+
+### Validation process
+
+1. Validator in `running` state will pull binlog from upstream and get `row change`s.
+    - validator will only validate `row change` which has already migrated by syncer.
+2. After routing and filtering using the same rule as in syncer, `row change` is dispatched to validator `worker`. `row change` of same table and primary key will be dispatched to the same `worker`.
+3. `worker` merges `row change` by table and primary key, we will use the last `row change` meet since last change override previous one. Then put it into `pending row changes`.
+4. After accumulates enough `row change`s or after some interval, `worker` queries downstream for data using table and primary key info in those `row change`s, then compare with data in `row change`s
+    - for `insert`/`update` type of row change, we validate them differently by validation mode
+      - in `full` validation mode, we compare them column by column
+      - in `fast` validation mode, we only check it's existence
+    - for `delete` type of `row change`, downstream should not contains that data.
+5. For `row change`s which are validated successfully, worker will remove it from `pending row changes`, others which fail to validate will be marked as `failed row change` and be validated again after some interval.
+6. If a `failed row change` keeps failing to validate for enough time(`delay time`), we mark it as `error row change` and save it into meta database.
+
+### False positive
+
+`error row change` produced in validation process might not be data which is incorrectly migrated, there are cases where `row change` is marked as falsely. Suppose some row which keeps changing on upstream for a time period > `delay time`. If it's marked as `failed row change` since the first time it changes, validator may mark it as `error row change` falsely. In real world scenarios, it's not common.
+
+To reduce the chance of false positive, validator only starts marking `failed row change` when validator has reached the progress of syncer or after some `initial delay time`
+
+
+### Validation checkpoint
+
+Validator will start validation from previous location after failover or resuming. Validator will save current location and current `pending row changes` into meta-db after some interval.
+
+### Self-protection
+
+Validator `worker` will cache `pending row changes` in memory, to avoid potential OOM, we add an self-protect mechaniam. If there's too much `pending row changes` or the overall size of `pending row changes` is too large, validator will be stopped automatically.
```

Comment:
```suggestion
Validator `worker` will cache `pending row changes` in memory, to avoid potential OOM, we add a self-protect mechanism. If there are too many `pending row changes` or the overall size of `pending row changes` is too large, validator will be stopped automatically.
```